win32/devdoc/threadpool_win32.md

```diff
@@ -243,3 +264,54 @@ static VOID CALLBACK on_timer_callback(PTP_CALLBACK_INSTANCE instance, PVOID con
 **SRS_THREADPOOL_WIN32_42_017: [** Otherwise `context` shall be used as the context created in `threadpool_schedule_work`. **]**
 
 **SRS_THREADPOOL_WIN32_42_018: [** The `work_function` callback passed to `threadpool_schedule_work` shall be called, passing to it the `work_function_context` argument passed to `threadpool_schedule_work`. **]**
+
+### threadpool_create_work_item
+
+```c
+MOCKABLE_FUNCTION(, THREADPOOL_WORK_ITEM_HANDLE, threadpool_create_work_item, THANDLE(THREADPOOL), threadpool, THREADPOOL_WORK_FUNCTION, work_function, void*, work_function_context);
+```
+
+`threadpool_create_work_item` creates a work item to be executed by the threadpool.
+
+**S_R_S_THREADPOOL_WIN32_05_004: [** If `threadpool` is `NULL`, `threadpool_create_work_item` shall fail and return a `NULL` value. **]**
+
+**S_R_S_THREADPOOL_WIN32_05_005: [** If `work_function` is `NULL`, `threadpool_create_work_item` shall fail and return a `NULL` value. **]**
+
+**S_R_S_THREADPOOL_WIN32_05_006: [** Otherwise `threadpool_create_work_item` shall allocate a context `work_item_context` of type `THREADPOOL_WORK_ITEM_HANDLE` where `work_function`, `work_function_context`, and `ptp_work` shall be saved. **]**
```

Comment:
>shall allocate a context `work_item_context` [](http://example.com/codeflow?start=77&length=44)

Can we avoid this allocation alltogether?

Reply:
It happens once and the function return type is THREADPOOL_WORK_ITEM_HANDLE. Or we can ask the caller to allocate it and it can be statically defined on their end? We can also have a double pointer and allocate only if it's content is a NULL pointer.

Reply:
I was hinting at whether we can cast PTP_WORK to our handle. Since both are pointers and I don't think we need to store anything else in our "work item context"

Reply:
In WIN32, this context gets passed to `on_work_callback` where it calls `context->work_function(work_function_context)`.

Reply:
Ok.